server/src/main/java/org/opensearch/ingest/AbstractBatchingProcessor.java

```diff
@@ -0,0 +1,141 @@
+        if (ingestDocumentWrappers.isEmpty()) {
+            handler.accept(Collections.emptyList());
+            return;
+        }
+
+        // if batch size is larger than document size, send one batch
+        if (this.batchSize >= ingestDocumentWrappers.size()) {
+            subBatchExecute(ingestDocumentWrappers, handler);
+            return;
+        }
+
+        // split documents into multiple batches and send each batch to batch processors
+        List<List<IngestDocumentWrapper>> batches = cutBatches(ingestDocumentWrappers);
+        int size = ingestDocumentWrappers.size();
+        AtomicInteger counter = new AtomicInteger(size);
+        List<IngestDocumentWrapper> allResults = Collections.synchronizedList(new ArrayList<>());
+        for (List<IngestDocumentWrapper> batch : batches) {
+            this.subBatchExecute(batch, batchResults -> {
+                allResults.addAll(batchResults);
+                if (counter.addAndGet(-batchResults.size()) == 0) {
+                    handler.accept(allResults);
+                }
+                assert counter.get() >= 0 : "counter is negative";
+            });
+        }
+    }
+
+    private List<List<IngestDocumentWrapper>> cutBatches(List<IngestDocumentWrapper> ingestDocumentWrappers) {
+        List<List<IngestDocumentWrapper>> batches = new ArrayList<>();
+        for (int i = 0; i < ingestDocumentWrappers.size(); i += this.batchSize) {
+            batches.add(ingestDocumentWrappers.subList(i, Math.min(i + this.batchSize, ingestDocumentWrappers.size())));
+        }
+        return batches;
+    }
+
+    /**
+     * Factory class for creating {@link AbstractBatchingProcessor} instances.
+     *
+     * @opensearch.internal
+     */
+    public abstract static class Factory implements Processor.Factory {
+        final String processorType;
+
+        protected Factory(String processorType) {
+            this.processorType = processorType;
+        }
+
+        /**
+         * Creates a new processor instance.
+         *
+         * @param processorFactories The processor factories.
+         * @param tag                 The processor tag.
+         * @param description         The processor description.
+         * @param config              The processor configuration.
+         * @return The new AbstractBatchProcessor instance.
+         * @throws Exception If the processor could not be created.
+         */
+        @Override
+        public AbstractBatchingProcessor create(
+            Map<String, Processor.Factory> processorFactories,
+            String tag,
+            String description,
+            Map<String, Object> config
+        ) throws Exception {
+            int batchSize = ConfigurationUtils.readIntProperty(this.processorType, tag, config, BATCH_SIZE_FIELD, DEFAULT_BATCH_SIZE);
+            if (batchSize < 1) {
+                throw newConfigurationException(
```

Comment:
Could we add a test case for `batchSize<1`?

Reply:
+1

Also, what about zero?

Reply:
Thanks for catching this, will add UT for exception situation.

@andrross zero should be an invalid value of batch_size